### PR TITLE
Fix gRPC channel pooling

### PR DESCRIPTION
**Pull request by ggezer**
ChannelKey's equals() method was wrong and was missing hashcode() implementation. This caused gRPC channel pooling to always create new managed channel upon every request. This PR fixes it and adds further optimization to delay actual channel destruction.

### CODE REVIEW

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java [hunk 1]
```diff
@@ -86,32 +146,21 @@ public ManagedChannel acquireManagedChannel(ChannelKey channelKey) {
...
-                  Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT),
-                  TimeUnit.MILLISECONDS);
-            } catch (InterruptedException e) {
-              Thread.currentThread().interrupt();
-              // Allow thread to exit.
-            } finally {
-              channel.shutdownNow();
-            }
-            Verify.verify(channel.isShutdown());
+            // Mark channel for destruction.
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: What was the issue with destroying here (or in other words, how does async destruction help things)?

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java [hunk 2]
```diff
@@ -50,12 +61,61 @@ public static GrpcManagedChannelPool INSTANCE() {
...
+        if (!channelRef.dereference()) {
+          iReactivated++;
+        } else {
+          iDestroyed++;
+          ManagedChannel channel = channelRef.reference();
+          channel.shutdown();
+          try {
+            channel.awaitTermination(
+                Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT),
+                TimeUnit.MILLISECONDS);
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: It seems inefficient to wait for channel termination while holding the resource lock? Ideally we can just remove the channel from `mChannels` and release the lock.

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java (hunk 1 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: Just an optimization. This was not the main issue. 
Delaying the actual channel destruction helps for concurrent clients to grab and activate inactive channels. Otherwise even in a highly concurrent scenario it'd be fairly often for the channel to get destroyed and recreated.

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java [hunk 3]
```diff
@@ -85,18 +82,14 @@ public GrpcManagedChannelPool() {
...
-      for (ChannelKey channelKey : mChannelsToRelease) {
-        Verify.verify(mChannels.containsKey(channelKey));
-        ManagedChannelReference channelRef = mChannels.get(channelKey);
-        channelRef.reference();
-        if (!channelRef.dereference()) {
-          iReactivated++;
-        } else {
-          iDestroyed++;
-          ManagedChannel channel = channelRef.reference();
+      int iDestroyed = 0;
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: We usually do not prefix with `i`

**Comment by ggezer** [comment]
@calvinjia working on adding a unit test.

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java (hunk 2 above)
Comment (calvinjia) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: I don't think this comment is addressed yet. Is there any way someone can get a reference to the channel after it has been removed from `mChannels`? If not, I think we can just remove from `mChannels` while holding the lock and then destroy the channels afterward?

Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: It is already deleted from the mChannels after successful destruction.
However channels that failed to destruct should not be served again so moving the deletion to before attempting the destruction.

Is this good enough?

Comment (calvinjia) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: My point is we are holding the exclusive lock while we wait for the channels to be destroyed. That seems unnecessary since we logically destroy the channel by removing it from the map (no one else can see the channel anymore), which is cheaper than waiting for termination. Once we have logically removed all the channels we want to destroy later, we can relinquish the lock since we have exclusive access to the channels we want to destroy.

Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: Makes sense. Will do.

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java [hunk 4]
```diff
@@ -227,7 +258,7 @@ private ManagedChannel get(){
    */
   public static class ChannelKey {
     private SocketAddress mAddress;
-    private boolean mPlain = false;
+    private boolean mPlain = true;
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: Why is this changed?

Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: I thought plain should be true as default in OS code base as it's the case.

File: core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java [hunk 5]
```diff
@@ -83,33 +107,40 @@ public GrpcManagedChannelPool() {
...
         }
       }
-      LOG.debug("gRPC channel terminator shut down {} channels. Remaining active channel count: {}",
-          iDestroyed, mChannels.size());
     }
+
+    for (Pair<ChannelKey, ManagedChannelReference> channelPair : channelsToDestroy) {
+      ManagedChannel channel = channelPair.getSecond().get();
+      channel.shutdown();
+      try {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: Consider calling shutdown on all the channels first then calling await on all of them. That should make the execution parallel.

Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: One of the reasons behind pooling is to serialize channel destruction as I was issues when channels are destructed concurrently. I'd like to postpone this as later improvement task. It's not critical as we don't expect many unique channels per process.

Comment (calvinjia) [reply core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java]: Let's leave a TODO then
